### PR TITLE
gettext: fix compiling with clang

### LINKED ISSUE #348658: Build failure: pkgsLLVM.gettext
### Steps To Reproduce

1. build pkgsLLVM.gettext

### Build log

```
./dcigettext.c:147:7: error: redeclaration of 'getcwd' must have the 'overloadable' attribute
  147 | char *getcwd ();
      |       ^
```

### Additional context

LLVM 19 caused this

### Notify maintainers

@jsierles zimbatm

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by RossComputerGuy**
## Things done

Fixes #348658

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by zimbatm — APPROVED** [review]
Sounds good. Maybe add a reference to the issue in the code for context.

**Comment by RossComputerGuy** [comment]
@zimbatm Done, I also changed it for Darwin so it doesn't need to go to staging.

**Comment by RossComputerGuy** [comment]
Rebuild 0 all around on Ofborg

File: pkgs/development/libraries/gettext/default.nix [hunk 1]
```diff
@@ -47,6 +47,10 @@ stdenv.mkDerivation rec {
     sed -i -e "s/\(libgettextsrc_la_LDFLAGS = \)/\\1..\/gnulib-lib\/libxml_rpl.la /" gettext-tools/src/Makefile.in
   '' + lib.optionalString stdenv.hostPlatform.isMinGW ''
     sed -i "s/@GNULIB_CLOSE@/1/" */*/unistd.in.h
+  ''
+    # PR: https://github.com/NixOS/nixpkgs/issues/348658
```
Comment (alyssais) [inline pkgs/development/libraries/gettext/default.nix:51]: This isn't a lot of explanation for a very confusing fix…

Do we know why it's being redeclared?

Comment (RossComputerGuy) [reply pkgs/development/libraries/gettext/default.nix:51]: I spent like 15 minutes looking into it and got completely stumped, this is the best solution I was able to find lol. The macros for things like having the config, libc, and getcwd exists so theoretically it should just work.

Comment (Artturin) [reply pkgs/development/libraries/gettext/default.nix:51]:
```suggestion
    # ISSUE: https://github.com/NixOS/nixpkgs/issues/348658
```

Comment (RossComputerGuy) [reply pkgs/development/libraries/gettext/default.nix:51]: Typo lol, I'm so used to writing PR with Nixpkgs LLVM workspace.

**Comment by Artturin** [comment]
I found the proper fix by upstream https://github.com/autotools-mirror/gettext/commit/cb2c1486336462c8180f487221181ee798b0e73e

**Comment by RossComputerGuy** [comment]
Alright, best we can probably do is drop in a diff of that commit and then drop it in a future update. It looks like the gettext upgrade isn't in a mergable state and it'll probably miss the staging cycle.

**Comment by Artturin** [comment]
> Alright, best we can probably do is drop in a diff of that commit and then drop it in a future update. It looks like the gettext upgrade isn't in a mergable state and it'll probably miss the staging cycle.

Yeah and it's a core package

**Comment by RossComputerGuy** [comment]
Yeah... I've added the patch.

File: pkgs/development/libraries/gettext/default.nix [hunk 2]
```diff
@@ -19,7 +19,16 @@ stdenv.mkDerivation rec {
     # fix reproducibile output, in particular in the grub2 build
     # https://savannah.gnu.org/bugs/index.php?59658
     ./0001-msginit-Do-not-use-POT-Creation-Date.patch
-  ];
+  ]
+    # An accidental inclusion in https://marc.info/?l=glibc-alpha&m=150511271003225&w=2
```
Comment (Artturin) [inline pkgs/development/libraries/gettext/default.nix:23]: Searching the glibc commits I can't find that this set eas ever included and was just something that other distros maintainer accidentally included in their build

**Comment by RossComputerGuy** [comment]
Alright, it seems to be in a good state. Gonna let ofborg finish.

**Pull request merged**